primitives/src/crh/poseidon/batched_crh.rs

```diff
@@ -0,0 +1,548 @@
+extern crate rand;
+extern crate rayon;
+
+use algebra::fields::mnt6753::Fr as MNT6753Fr;
+use algebra::fields::mnt4753::Fr as MNT4753Fr;
+use algebra::{PrimeField, MulShort};
+
+use std::marker::PhantomData;
+
+use crate::crh::poseidon::{
+    parameters::{MNT4753PoseidonParameters, MNT6753PoseidonParameters}
+};
+
+use crate::crh::BatchFieldBasedHash;
+use crate::{Error, PoseidonParameters, matrix_mix_short};
+
+
+pub struct PoseidonBatchHash<F: PrimeField, P: PoseidonParameters<Fr = F>>{
+    _field:      PhantomData<F>,
+    _parameters: PhantomData<P>,
+}
+
+impl<F: PrimeField + MulShort, P: PoseidonParameters<Fr=F>> PoseidonBatchHash<F, P> {
+
+    fn poseidon_full_round(vec_state: &mut [Vec<P::Fr>], round_cst_idx: &mut usize) {
+
+        // For each of the element position of the state vector
+        for j in 0..P::T {
+
+            // get the constant associated to element position of the state vector
+            let rc = P::ROUND_CST[*round_cst_idx];
+
+            // go over each of the state vectors and add the constant
+            for k in 0..vec_state.len() {
+                vec_state[k][j] += &rc;
+            }
+            *round_cst_idx += 1;
+        }
+
+        // Apply the S-BOX to each of the elements of the state vector
+        // Use Montgomery simultaneous inversion
+        let mut w: Vec<P::Fr> = Vec::new();
+        let mut accum_prod = P::Fr::one();
+
+        w.push(accum_prod);
+        // Calculate the intermediate partial products
+        for i in 0..vec_state.len() {
+            for j in 0..P::T {
+                accum_prod = accum_prod * &vec_state[i][j];
+                w.push(accum_prod);
+            }
+        }
+
+        // if the accum_prod is zero, it means that one of the S-Boxes is zero
+        // in that case compute the inverses individually
+        if accum_prod == P::Fr::zero() {
+            for i in 0..vec_state.len() {
+                for j in 0..P::T {
+                    if vec_state[i][j] != P::Fr::zero() {
+                        vec_state[i][j] = vec_state[i][j].inverse().unwrap();
+                    }
+                }
+            }
+        } else {
+
+            // Calculate the inversion of the products
+            // The inverse always exists in this case
+            let mut w_bar = accum_prod.inverse().unwrap();
+
+            // Extract the individual inversions
```

Comment:
Is this actually cheaper than computing the inversions individually (like at lines 57-63) ?